projects/plugins/jetpack/extensions/shared/execution-lock.js

```diff
@@ -0,0 +1,74 @@
+/**
+ * Map that relates a locked key with its unique lock value.
+ *
+ * @type {Map<string, number>}
+ */
+const lockMap = new Map();
+
+/**
+ * Acquires a lock for a given key.
+ *
+ * @param {string} key - Key for which we will acquire a lock.
+ * @returns {null|[string,number]} - After successfully acquiring a lock we return the lock, if the key is locked, null is returned.
+ */
+const acquire = key => {
+	if ( isLocked( key ) ) {
+		return null;
+	}
+
+	const lockValue = Math.random();
+	lockMap.set( key, lockValue );
+	return [ key, lockValue ];
+};
+
+/**
+ * Block the execution until the locked key is released without blocking the thread.
+ *
+ * @param {string} key - Key that we will wait for release.
+ * @param {number} timeOffset - Amount of time to wait between lock checks.
+ */
+const blockExecution = async ( key, timeOffset = 10 ) => {
+	if ( isLocked( key ) ) {
+		await new Promise( resolve => setTimeout( resolve, timeOffset ) );
+		await blockExecution( key, timeOffset );
+	}
+};
+
+/**
+ * Clears all the created locks.
+ *
+ * @returns {void}
+ */
+const clearAll = () => lockMap.clear();
+
+/**
+ * Checks if a lock has been acquired for the given key.
+ *
+ * @param {string} key - Key to check
+ * @returns {boolean} - True if a lock has been acquired for the given key, false otherwise.
+ */
+const isLocked = key => lockMap.has( key );
+
+/**
+ * Releases a given lock.
+ *
+ * @param {[string,number]} lock - Tuple containing the key and lock value obtained by invoking the acquire method.
+ * @returns {boolean} - True if the lock is released, false otherwise.
+ */
+const release = lock => {
+	const [ key, lockedBy ] = lock;
+	if ( isLocked( key ) && lockMap.get( key ) === lockedBy ) {
+		lockMap.delete( key );
+		return true;
+	}
+
+	return false;
+};
+
+export default {
```

Comment:
Key-based lock implementation, the only special feature is the block execution method, which prevents some code from being executed without necessarily blocking the main thread. Implemented as a singleton.

Reply:
> I know that this might a bit too late since we already have an implementation. Since by definition having a lock means having a global state, could we change the implementation to avoid making the additional API requests? Right now, the lock just delays the status requests until the first one is finished.

I've thought of this too, I was going to recap some things I've found in a P2. We can't avoid making those requests (at least without adding yet more code specific to it) because of the way the Store works. If I'm not mistaken, when you invoke the getProducts selector if a resolver with the same name exists, the store will return the current state and call the resolvers method asynchronously to update the data.
[This article](https://unfoldingneurons.com/2020/wordpress-data-store-properties-resolvers) I've found explains how a typical store works.